Add support for generating a consumer for the CSS property 'container-name'

weinig opened the pull request:
#### 1022bf35b113863cf362474a7d6fd2b813b3ebfc
<pre>
Add support for generating a consumer for the CSS property &apos;container-name&apos;
<a href="https://bugs.webkit.org/show_bug.cgi?id=288380">https://bugs.webkit.org/show_bug.cgi?id=288380</a>

Reviewed by Darin Adler.

Adds enough support in process-css-properties.py for generating a consumer
for the CSS property &apos;container-name&apos;.

This requires:
   - adding syntax to define a list of excluded values for &lt;custom-ident&gt;:
       e.g. &lt;custom-ident excluding=none,auto&gt;
   - adding generation support for the &apos;+&apos; multiplier.

This also allowed for generating &apos;view-transition-class&apos;, which has a very
similar grammar.

To add support for the excluding syntax, the existing term attributes such
as &apos;unitless-allowed&apos; in &lt;length unitless-allowed&gt; were updated to use the
same syntax as excluding, making it now, &apos;unitless=allowed&apos;.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Conditional.cpp: Removed.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Conditional.h: Removed.
    - Removes no longer needed CSSPropertyParserConsumer+Conditional.h/cpp

* Source/WebCore/css/process-css-properties.py:
    - Overhauls reference term attribute processing to support a new syntax for
      attributes: &lt;my-term foo=bar,baz,...&gt;
    - Support for range attributes remains.
    - Builtin processing of those attributes has changed to use the name of the
      attribute (the &apos;foo&apos; in &apos;foo=bar,baz,...&apos;) as the descriptor name and now
      only a single range attribute is supported, which is necessary as these don&apos;t
      have a name associated with them.
    - Range attribute also store their min/max values faithfully, rather than
      eagerly stringifying them. This will allow us to use the values for parameters
      in the future.

* Source/WebCore/css/CSSProperties.json:
    - Update grammars that use &lt;custom-ident&gt; and need to exclude terms to use
      the new excluding= syntax.
    - Update grammars that use the old production parameter syntax using a single
      string to use the new foo=bar syntax. (e.g. unitless-allowed -&gt; unitless=allowed)
    - Update parser-grammar-unused-comments to remove now supported productions.
    - Export &lt;single-container-name&gt; so it can be used by the container query parser.

* Source/WebCore/css/query/ContainerQueryParser.cpp:
(WebCore::CQ::ContainerQueryParser::consumeContainerQuery):
    - Use newly generated and exported CSSPropertyParsing::consumeSingleContainerName.

* Source/WebCore/css/parser/CSSParserIdioms.h:
* Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
    - Move isValidContainerNameIdentifier to CSSParserIdioms and make it also
      do the custom-ident check, simplifying caller.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::consumeContainerShorthand):
    - Use parseSingleValue and pass the current shorthand to ensure expectations
      of shorthand parsing is followed, and allowing us not to export the container-name
      consumer function.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Ident.h:
(WebCore::CSSPropertyParserHelpers::consumeCustomIdentExcluding):
    - Add new consumer for custom-idents that takes a list of CSSValueIDs to disallow.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+List.h:
(WebCore::CSSPropertyParserHelpers::consumeListSeparatedBy):
    - Adds support for consuming space separated lists.
    - Adds support for indicating minimum and maximum list length constraints. By default,
      all list consumers have no constraints applied.
    - Refactor list consumers into a single template function.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+ViewTransition.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ViewTransition.h:
    - Remove now unneeded hand written consumer for &apos;view-transition-class&apos;.

Canonical link: <a href="https://commits.webkit.org/291086@main">https://commits.webkit.org/291086@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0d409adf58795144363d3f97f457ce6a509c7d50

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91974 "30 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11506 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/1052 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/96904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/42573 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/94024 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/11847 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19993 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/96904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/42573 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94975 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/11847 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/83302 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/96904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/11847 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/902 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/41788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/11847 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/897 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98936 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/19090 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/14113 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/79602 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/19342 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79153 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/78830 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19514 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/676 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/12128 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/19071 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/24280 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/18768 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/22227 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/20519 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Ident.h:
```diff
@@ -106,6 +106,12 @@ std::optional<typename Map::ValueType> peekIdentUsingMapping(CSSParserTokenRange
+    if (std::ranges::any_of(excluding, [peekedValueID = range.peek().id()](const auto& valueID) { return valueID == peekedValueID; }))
+        return nullptr;
+    return consumeCustomIdent(range, shouldLowercase);
+}
```
I slightly prefer to have both a declaration and a separate definition, so the implementation is not interspersed with the interface. I haven’t looked over the whole file to see if that would enhance things here, but I’d be tempted to do it that way.

weinig replied:
The file is a pretty bad offender, but its not reason not to fix it up.

weinig committed:
Add support for generating a consumer for the CSS property 'container-name'
https://bugs.webkit.org/show_bug.cgi?id=288380

Reviewed by Darin Adler.

Adds enough support in process-css-properties.py for generating a consumer
for the CSS property 'container-name'.

This requires:
   - adding syntax to define a list of excluded values for <custom-ident>:
       e.g. <custom-ident excluding=none,auto>
   - adding generation support for the '+' multiplier.

This also allowed for generating 'view-transition-class', which has a very
similar grammar.

To add support for the excluding syntax, the existing term attributes such
as 'unitless-allowed' in <length unitless-allowed> were updated to use the
same syntax as excluding, making it now, 'unitless=allowed'.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Conditional.cpp: Removed.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Conditional.h: Removed.
    - Removes no longer needed CSSPropertyParserConsumer+Conditional.h/cpp

* Source/WebCore/css/process-css-properties.py:
    - Overhauls reference term attribute processing to support a new syntax for
      attributes: <my-term foo=bar,baz,...>
    - Support for range attributes remains.
    - Builtin processing of those attributes has changed to use the name of the
      attribute (the 'foo' in 'foo=bar,baz,...') as the descriptor name and now
      only a single range attribute is supported, which is necessary as these don't
      have a name associated with them.
    - Range attribute also store their min/max values faithfully, rather than
      eagerly stringifying them. This will allow us to use the values for parameters
      in the future.

* Source/WebCore/css/CSSProperties.json:
    - Update grammars that use <custom-ident> and need to exclude terms to use
      the new excluding= syntax.
    - Update grammars that use the old production parameter syntax using a single
      string to use the new foo=bar syntax. (e.g. unitless-allowed -> unitless=allowed)
    - Update parser-grammar-unused-comments to remove now supported productions.
    - Export <single-container-name> so it can be used by the container query parser.

* Source/WebCore/css/query/ContainerQueryParser.cpp:
(WebCore::CQ::ContainerQueryParser::consumeContainerQuery):
    - Use newly generated and exported CSSPropertyParsing::consumeSingleContainerName.

* Source/WebCore/css/parser/CSSParserIdioms.h:
* Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
    - Move isValidContainerNameIdentifier to CSSParserIdioms and make it also
      do the custom-ident check, simplifying caller.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::consumeContainerShorthand):
    - Use parseSingleValue and pass the current shorthand to ensure expectations
      of shorthand parsing is followed, and allowing us not to export the container-name
      consumer function.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Ident.h:
(WebCore::CSSPropertyParserHelpers::consumeCustomIdentExcluding):
    - Add new consumer for custom-idents that takes a list of CSSValueIDs to disallow.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+List.h:
(WebCore::CSSPropertyParserHelpers::consumeListSeparatedBy):
    - Adds support for consuming space separated lists.
    - Adds support for indicating minimum and maximum list length constraints. By default,
      all list consumers have no constraints applied.
    - Refactor list consumers into a single template function.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+ViewTransition.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ViewTransition.h:
    - Remove now unneeded hand written consumer for 'view-transition-class'.

Canonical link: https://commits.webkit.org/291086@main
